src/SoftGlobalScope.jl

```diff
@@ -89,15 +89,17 @@ else
             return Expr(:try, try_clause, ex.args[2], catch_clause, finally_clause)
         elseif isexpr(ex, :let)
             letglobals = setdiff(globals, localvars(ex.args[1]))
-            return Expr(ex.head, _softscope(ex.args[1], globals, insertglobal),
-                                _softscope(ex.args[2], letglobals, true))
+            return Expr(ex.head, ex.args[1], _softscope(ex.args[2], letglobals, true))
         elseif isexpr(ex, :block) || isexpr(ex, :if)
             return Expr(ex.head, _softscope.(ex.args, Ref(globals), insertglobal)...)
         elseif isexpr(ex, :local)
             setdiff!(globals, localvars(ex.args)) # affects globals in surrounding scope!
             return ex
         elseif insertglobal && ex.head in assignments && ex.args[1] in globals
             return Expr(:global, Expr(ex.head, ex.args[1], _softscope(ex.args[2], globals, insertglobal)))
+        elseif !insertglobal && isexpr(ex, :(=))
```

Comment:
Won't this treat `x` as global even if it has been explicitly declared as `local`?

Reply:
We probably need a set `locals`, in addition to `globals`, to track local variables.

Reply:
No, the `elseif isexpr(ex, :local)` beforehand catches that one
```
julia> a = 1
1

julia> softscope(Main, :(begin; local a = 2; a += 1; end))
quote
    #= REPL[4]:1 =#
    local a = 2
    #= REPL[4]:1 =#
    a += 1
end
```

Reply:
Ah, though subsequent assignments will get caught... OK, so we might need a separate list of locals.

Reply:
Sorry didn't see your extra comments - browser didn't refresh.

Your first example is problematic but the second example is fine with this PR. In the second example, the let introduces a local scope and so `insertglobals` is true -> the assignment is not recorded.